File: solutions/java/say/1/src/main/java/Say.java
public class Say {

    public String subTwenty(long number){
        if (number == 19) return "nineteen";
        else if (number == 18) return "eighteen";
        else if (number == 17) return "seventeen";
        else if (number == 16) return "sixteen";
        else if (number == 15) return "fifteen";
        else if (number == 14) return "fourteen";
        else if (number == 13) return "thirteen";
        else if (number == 12) return "twelve";
        else if (number == 11) return "eleven";
        else if (number == 10) return "ten";
        else if (number == 9) return "nine";
        else if (number == 8) return "eight";
        else if (number == 7) return "seven";
        else if (number == 6) return "six";
        else if (number == 5) return "five";
        else if (number == 4) return "four";
        else if (number == 3) return "three";
        else if (number == 2) return "two";
        else if (number == 1) return "one";
        return "";
    }

    public String subThousand(long number){
        String result = "";
        if (number < 1_000L && number >= 100L) {
            long hundreds = number / 100L;
            result += subTwenty(hundreds) + " hundred";
            number = number % 100L;

            if (number > 0L) result += " ";
        } 
        
        if (number < 100L && number >= 20L) {
            long tens = number / 10L;
            
            if (tens == 9L) result += "ninty";
            if (tens == 8L) result += "eighty";
            if (tens == 7L) result += "seventy";
            if (tens == 6L) result += "sixty";
            if (tens == 5L) result += "fifty";
            if (tens == 4L) result += "forty";
            if (tens == 3L) result += "thirty";
            if (tens == 2L) result += "twenty";

            number = number % 10L;

            if (number > 0L) result += "-";
        }
        
        if (number <= 20L){
            result += subTwenty(number);
            number = 0L;
        }

        return result;
    }

    public String say(long number) {
        if (number < 0L) throw new java.lang.IllegalArgumentException();
        if (number >= 1_000_000_000_000L) throw new java.lang.IllegalArgumentException();
        
        if (number == 0L) return "zero";

        String result = "";

        if (number >= 1_000_000_000L) {
            long billions = number / 1_000_000_000L;
            result += subThousand(billions) + " billion";
            number %= 1_000_000_000L;
            if (number > 0L) result += " ";
        }
        
        if (number >= 1_000_000L) {
            long millions = number / 1_000_000L;
            result += subThousand(millions) + " million";
            number %= 1_000_000L;
            if (number > 0L) result += " ";
        }

        if (number >= 1_000L) {
            long thousands = number / 1_000L;
            result += subThousand(thousands) + " thousand";
            number %= 1_000L;
            if (number > 0L) result += " ";
        }  

        if (number > 0L){
            result += subThousand(number);
            number = 0L;
        }

        return result;
    }
}

File: solutions/python/say/1/say.py
def say_to_nineteen(number):
    words = [
        "zero",
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
        "ten",
        "eleven",
        "twelve",
        "thirteen",
        "fourteen",
        "fifteen",
        "sixteen",
        "seventeen",
        "eighteen",
        "nineteen"
    ]
    return words[number]


def say_tens(number):
    tens_place = [None, None, "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]
    match (number):
        case n if 0 <= n <= 19:
            return say_to_nineteen(number)
        case n if n % 10 == 0:
            return tens_place[number // 10]
        case _:
            return f"{tens_place[number // 10]}-{say_to_nineteen(number % 10)}"

def say_hundreds(number):
    match (number):
        case n if 0 <= n <= 99:
            return say_tens(number)
        case n if 100 <= n <= 999 and n % 100 == 0:
            return f"{say_to_nineteen((number // 100) % 10)} hundred"
        case _:
            return f"{say_to_nineteen((number // 100) % 10)} hundred {say_tens(number % 100)}"

def say_thousands(number):
    match (number):
        case n if 0 <= n <= 999:
            return say_hundreds(number)
        case n if 1000 <= n <= 999_999 and n % 1_000 == 0:
            return f"{say_hundreds((number // 1_000) % 1000)} thousand"
        case _:
            return f"{say_hundreds((number // 1_000) % 1000)} thousand {say_hundreds(number % 1000)}"

def say_millions(number):
    match (number):
        case n if 0 <= n <= 999_999:
            return say_thousands(number)
        case n if 1_000_000 <= n <= 999_999_999 and n % 1_000_000 == 0:
            return f"{say_hundreds((number // 1_000_000) % 1000)} million"
        case _:
            return f"{say_hundreds((number // 1_000_000) % 1000)} million {say_thousands(number % 1_000_000)}"

def say_billions(number):
    match (number):
        case n if 0 <= n <= 999_999_999:
            return say_millions(number)
        case n if 1_000_000_000 <= n <= 999_999_999_999 and n % 1_000_000_000 == 0:
            return f"{say_hundreds((number // 1_000_000_000) % 1000)} billion"
        case _:
            return f"{say_hundreds((number // 1_000_000_000) % 1000)} billion {say_millions(number % 1_000_000_000)}"

def say(number):
    match (number):
        case n if 0 <= n <= 999_999_999_999:
            return say_billions(number)
        case _:
            raise ValueError("input out of range")

